crates/lair_keystore_api/src/lair_client/async_io.rs

```diff
@@ -0,0 +1,163 @@
+//! Wrap a raw tokio::io::Async{Read, Write} channel into a LairClient.
+
+use super::*;
+
+/// Wrap a raw tokio::io::Async{Read, Write} channel into a LairClient.
+pub fn new_async_io_lair_client<S, R>(
+    send: S,
+    recv: R,
+) -> impl Future<Output = LairResult<LairClient>> + 'static + Send
+where
+    S: tokio::io::AsyncWrite + 'static + Send + Unpin,
+    R: tokio::io::AsyncRead + 'static + Send + Unpin,
+{
+    async move {
+        // wrap the channels in sodium_secretstream
+        let (send, recv) =
+            crate::sodium_secretstream::new_s3_pair::<LairApiEnum, _, _>(
+                send, recv, false,
+            )
+            .await?;
+
+        // derive our encryption (to server) secret context key
+        let enc_ctx_key = <sodoken::BufWriteSized<32>>::new_mem_locked()?;
+        sodoken::kdf::derive_from_key(
+            enc_ctx_key.clone(),
+            142,
+            *b"ToSrvCxK",
+            send.get_enc_ctx_key(),
+        )?;
+        let enc_ctx_key = enc_ctx_key.to_read_sized();
+
+        // derive our decryption (from server) secret context key
+        let dec_ctx_key = <sodoken::BufWriteSized<32>>::new_mem_locked()?;
+        sodoken::kdf::derive_from_key(
+            dec_ctx_key.clone(),
+            42,
+            *b"ToCliCxK",
+            send.get_dec_ctx_key(),
+        )?;
+        let dec_ctx_key = dec_ctx_key.to_read_sized();
+
+        // build up our inner item
+        let inner = CliInner {
+            enc_ctx_key,
+            dec_ctx_key,
+            send: send.clone(),
+            pending: HashMap::new(),
+        };
+
+        let inner = Arc::new(RwLock::new(inner));
+
+        {
+            // spawn a task to manage incoming data
+            let inner = inner.clone();
+            tokio::task::spawn(async move {
+                let inner = &inner;
+                let send = &send;
+
+                recv.for_each_concurrent(4096, move |incoming| async move {
```

Comment:
I'm not sure if this is relevant to lair but I've found in practice going above about 100 concurrent futures ends up starving the futures in holochain but it could be dependant on the type of futures maybe.

Reply:
if these get paused, they largely get paused on awaiting a oneshot, so there shouldn't be much cpu demand. That being said, 4096 may be excessive, essentially this is how many open/pending requests a single connection can have... but we can always tweak this going forward.